File: 品味Java的21種設計模式/dp_java/iterator/example7/Iterator.java
package cn.javass.dp.iterator.example7;
/**
 * 迭代器接口，定义访问和遍历元素的操作，实现双向迭代
 */
public interface Iterator {
	/**
	 * 移动到聚合对象的第一个位置
	 */
	public void first();
	/**
	 * 移动到聚合对象的下一个位置
	 */
	public void next();
	/**
	 * 判断是否已经移动聚合对象的最后一个位置
	 * @return true表示已经移动聚合对象的最后一个位置，
	 *         false表示还没有移动到聚合对象的最后一个位置
	 */
	public boolean isDone();
	/**
	 * 获取迭代的当前元素
	 * @return 迭代的当前元素
	 */
	public Object currentItem();
	
	
	/**
	 * 判断是否为第一个元素
	 * @return 如果为第一个元素，返回true，否则返回false
	 */
	public boolean isFirst();
	/**
	 * 移动到聚合对象的上一个位置
	 */
	public void previous();
}

File: 23種設計模式及對應C++示例代碼/行为模式/Iterator_exam/Iterator.cpp
// Iterator.cpp: implementation of the Iterator class.
//
//////////////////////////////////////////////////////////////////////

#include "stdafx.h"
#include "Iterator.h"

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

#include "Aggregate.h"
#include <iostream>
using namespace std;

Iterator::Iterator(){}
Iterator::~Iterator(){}

ConcreteIterator::ConcreteIterator(Aggregate* ag , int idx)
{
	this->_ag = ag;
	this->_idx = idx;
}
ConcreteIterator::~ConcreteIterator() {}

Object ConcreteIterator::CurrentItem()
{
	return _ag->GetItem(_idx);
}
void ConcreteIterator::First()
{
	_idx = 0;
}

void ConcreteIterator::Next()
{
	if (_idx < _ag->GetSize()) _idx++;
}

bool ConcreteIterator::IsDone()
{
	return (_idx == _ag->GetSize());
}
